pkg/scheduler/cron.go

```diff
@@ -14,45 +14,168 @@
 package scheduler
 
 import (
+	"context"
+	"sync"
 	"time"
 
+	"github.com/go-logr/zapr"
+	"github.com/pingcap/log"
+	perr "github.com/pkg/errors"
 	cron "github.com/robfig/cron/v3"
+	"go.uber.org/zap"
 
 	"github.com/chaos-mesh/chaosd/pkg/core"
 )
 
 type Scheduler struct {
 	*cron.Cron
+	expStore    core.ExperimentStore
+	expRunStore core.ExperimentRunStore
+	cronStore   CronStore
 }
-
 type CronJob struct {
-	experiment core.Experiment
-	run        func()
+	// immutable fields
+	scheduler   *Scheduler
+	experiment  *core.Experiment
+	attackFunc  func() error
+	recoverFunc func() error
+
+	// mutable fields protected by sync.Locker
+	waitForRecovery bool
+	lock            sync.Mutex
+}
+
+func hasCronDurationExceeded(startedAt time.Time, duration time.Duration) bool {
+	return time.Until(startedAt.Add(duration)).Milliseconds() < 0
+}
+
+func (cj *CronJob) RecoverRun(expRun *core.ExperimentRun) {
+	defer cj.setWaitForRecovery(false)
+	log.Info("recovering attack on exp run", zap.String("expRunUID", expRun.UID))
+	if err := cj.recoverFunc(); err != nil {
+		log.Warn("recovery failed", zap.Error(err))
+	} else {
+		if err := cj.scheduler.expRunStore.Update(context.Background(), expRun.UID, core.RunRecovered, ""); err != nil {
+			log.Error("failed to update in DB", zap.Error(err))
+		}
+	}
+}
+
+func (cj *CronJob) assertNoRecovery() bool {
+	cj.lock.Lock()
+	waitingForRecovery := cj.waitForRecovery
+	cj.lock.Unlock()
+	return !waitingForRecovery
+}
+
+func (cj *CronJob) setWaitForRecovery(waitForRecovery bool) {
+	cj.lock.Lock()
+	cj.waitForRecovery = waitForRecovery
+	cj.lock.Unlock()
 }
 
-// TODO: on running, create new experiment run
 // TODO: write tests for it
-func (cj CronJob) Run() {
-	cj.run()
+func (cj *CronJob) Run() {
```

Comment:
seems that the `CronJob` is implementing the interface [Job](https://github.com/robfig/cron/blob/bc59245fe10efaed9d51b56900192527ed733435/cron.go#L35) how about adding a comment?

Reply:
sure, will add it

Reply:
fixed